### ISSUE TITLE
[css-conditional] [css-contain] Fleshing out @container queries with single-axis containment

### ISSUE THREAD

**Issue by mirisuzanne** [w3c/csswg-drafts#5796]
Update: As feedback comes in from various channels, I've been updating & maintaining a full [explainer with a changelog](https://github.com/oddbird/css-sandbox/blob/main/src/rwd/query/explainer.md).

## Intro

Media-queries allow an author to make style changes based on the overall viewport dimensions -- but in many cases, authors would prefer styling modular components based on their context within a layout. Earlier this year, David Baron & Brian Kardell proposed two complementary approaches to explore: the [`@container` rule][dbaron-cq], and a [`switch()`][switch] function. Both seem useful in different situations.

[dbaron-cq]: https://github.com/dbaron/container-queries-implementability
[switch]: https://bkardell.com/blog/AllThemSwitches.html

This proposal builds on David Baron's `@container` approach, which works by applying size & layout containment to the queried elements. Any element with both size & layout containment can be queried using a new `@container` rule, with similar syntax to existing media-queries. Currently, size containment is all-or-nothing. In order to make that less restrictive for authors, there are proposed single-axis (`inline-size` & `block-size`) values for the `contain` property.

This is a rough outline of the feature as I imagine it -- but there are a number of questions that could more easily resolved with a prototype. The purpose of this document is to flesh out some direction for more applied testing & exploration.

## Table of contents:
_Sadly you can't link sections in an issue…_

- Goals
- Non-goals
- Proposed Solutions
  - Single-axis containment (`inline-size` & `block-size` values)
  - Containment context
  - Container queries (`@container`)
- Key scenarios
  - Modular components in any container
  - Components with internal containers
  - Component in a responsive grid track
- Detailed design discussion & alternatives
  - Single-axis containment issues
  - Implicit vs explicit containers
  - Combining scope with container queries?
  - @-Rule or pseudo-class?
  - Questions to explore
- References & acknowledgements

## Goals

Often the layout of a page involves different "container" areas -- such as sidebars and main content areas -- with modular 
"component" parts that can be placed in any area. Those components can be complex (a full calendar widget) or fairly simple (basic typography), but should respond in some way to the size of the container.

This can happen at multiple levels of layout, even inside nested components. The important distinction is that the "container" is distinct from the "component" being styled. Authors can query one to style the other.

## Non-goals

Modern layouts provide a related problem with slightly different constraints. When using grid layout, for example, available grid-track size can change in ways that are difficult to describe based on viewport sizes -- such as shrinking & growing in regular intervals as new columns are added or removed from a repeating `auto-fit`/`auto-fill` grid. In this case there is no external "container" element to query for an accurate sense of available space.

There is a [reasonable workaround](#component-in-a-responsive-grid-track) in this proposal, but a more ideal solution might look more like Brian Kardell's [`switch()` proposal][switch]. It would be good to consider these approaches together, as they make complementary tradeoffs.

There are also proposed improvements to flexbox & grid -- such as [indefinite grid spans](https://github.com/w3c/csswg-drafts/issues/388) or [first / last keywords](https://github.com/w3c/csswg-drafts/issues/2402) -- which would help expand on the responsive nature of those tools.

And finally, authors often want to _smoothly interpolate values_ as the context changes, rather than toggling them at breakpoints. That would require a way to describe context-based animations, with breakpoint-like keyframes. Scott Kellum has been doing a lot of work in that area.

All of those would contribute towards a larger goal of "responsive components" in CSS. I don't think we can solve it all in a single feature.

### Single-axis containment (`inline-size` & `block-size` values)

_This is a proposed change to the [CSS Containment Module](https://drafts.csswg.org/css-contain/), specifically [size containment](https://drafts.csswg.org/css-contain/#containment-size) -- and already has [an issue thread for discussion](https://github.com/w3c/csswg-drafts/issues/1031)._

In order for container-queries to work in a performant way, authors will need to define container elements with explicit containment on their layout and queried-dimensions. This can be done with the existing `contain` property, using the `size` and `layout` values:

```css
.container {
  contain: size layout;
}
```

While that will work for some use-cases, the majority of web layout is managed through constraints on a _single (often inline) axis_. Intrinsic sizing on the cross (often block) axis is required to allow for changes in content, font size, etc. So this proposal would rely on one or two new single-axis values for `contain`, as discussed in #1031:

```css
.inline-container {
  contain: inline-size;
}

.block-container {
  contain: block-size;
}
```

Of these two values, it is clear that `block-size` has the fewer use-cases, and more potential implementation issues. Support for `block-size` would be great, but is not required to make container queries useful.

### Containment context

Ideally, container queries could be resolved against the _available space_ for any given element. Since size and layout containment are required, we instead need to define the _containment context_ for each element.

I'm proposing that any element with layout and size containment on a given axis generates a new _containment context_ in that axis, which descendants can query against:

```css
.two-axis-container {
  /* establishes a new containment context on both axis */
  contain: layout size;
}

.inline-container {
  /* establishes a new containment context on the inline axis */
  contain: layout inline-size;
}

.block-container {
  /* establishes a new containment context on the block axis */
  contain: layout block-size;
}
```

When size-containment is only available on a single axis, queries on the cross-axis will not resolve against that query context. When no containment context is established, there should be a fallback akin to the _Initial Containing Block_ which queries can resolve against.

### Container queries (`@container`)

This could be added to a future level of the [CSS Conditional Rules Module](https://drafts.csswg.org/css-conditional-3/). 

The `@container` rule can be used to style elements based on their immediate _containment context_, and uses a similar syntax to existing media queries. Where possible `@container` syntax should follow the established specifications for _conditional group rules_, and allow queries to be combined in a list:

```css
/* @container <container-query-list> { <stylesheet> } */
@container (width > 45em) {
  .media-object {
    grid-template: "img content" auto / auto 1fr;
  }
}
```

This would target any `.media-object` whose _containment context_ (nearest ancestor with containment applied) is greater-than `45em`. When no containment context is established, the _Initial Containing Block_ can be used to resolve the query.

Unlike media-queries, each element that is targeted by a conditional group rule will need to resolve the query against its own containment context. Different elements targeted by the same selector within the same query may still resolve differently based on context. Consider the following CSS & HTML together:

```css
/* css */
section {
  contain: layout inline-size;
}

div {
  background: red;
}

@container (width > 500px) {
  div {
    background: yellow;
  }
}

@container (width > 1000px) {
  div {
    background: green;
  }
}
```

```html
<!-- html -->
<section style="width: 1500px">
  <div>green background</div>

  <section style="width: 50%">
    <div>yellow background (resolves against inner section)</div>
  </section>
</section>
<section style="width: 400px">
  <div>red background</div>
</section>
```

Each `div` resolves differently based on its own immediate context.

#### Container features

Like media-queries, `@container` needs a well defined list of "features" that can be queried. The most essential container features are the contained dimensions:

- physical dimensions: `width` / `height`
- logical dimensions: `inline-size` / `block-size`

When containment is applied on both axis, we might also be able to query dimensional relationships such as:

- `aspect-ratio`
- `orientation`

Since container queries resolve against styled elements in the DOM, it may also be possible to query other aspects of the container's computed style?

- `inline-content-box`
- `font-size`
- etc.

This needs more discussion and fleshing-out.

### Modular components in any container

Page layouts often provide different layout "areas" that can act as containers for their descendant elements. These can be nested in more complex ways, but let's start with a sidebar and main content:

```html
<body>
  <main>...</main>
  <aside>...</aside>
</body>
```

We can establish a responsive layout, and declare each of these areas as a _containment context_ for responsive components:

```css
body {
  display: grid;
  grid-template: "main" auto "aside" auto / 100%;
}

@media (width > 40em) {
  body {
    grid-template: "aside main" auto / 1fr 3fr;
  }
}

main,
aside {
  contain: layout inline-size;
}
```

Now components can move cleanly between the two areas -- responding to container dimensions without concern for the overall layout. For example, some responsive defaults on typographic elements:

```css
h2 {
  font-size: 120%;
}

@container (width > 40em) {
  h2 {
    font-size: calc(130% + 0.5vw);
  }
}
```

Or "media objects" that respond to available space:

```css
.media-object {
  grid-template: "img" auto "content" auto / 100%;
}

@container (width > 45em) {
  .media-object {
    grid-template: "img content" auto / auto 1fr;
  }
}
```

### Components with internal containers

A more complex component, like a calendar, might reference external context while also defining nested containers:

```html
<section class="calendar">
  <div class="day">
    <article class="event">...</article>
    <article class="event">...</article>
  </div>
  <div class="day">...</div>
  <div class="day">...</div>
</section>
```

```css
.day {
  contain: layout inline-size;
}

/* despite having different containers, these could share a query */
@container (width > 40em) {
  /* queried against external page context */
  .calendar {
    grid-template: repeat(7, 1fr);
  }

  .day {
    border: thin solid silver;
    padding: 1em;
  }

  /* queried against the day */
  .event {
    grid-template: "img content" auto / auto 1fr;
  }
}
```

### Component in a responsive grid track

In some situations, there is no clear "container" element defining the available space. Consider the following HTML & CSS:

```html
<section class="card-grid">
  <div class="card">...</div>
  <div class="card">...</div>
  <div class="card">...</div>
  <div class="card">...</div>
</section>
```

```css
.card-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(20em, 1fr));
}

.card {
  display: grid;
  /* we want to change this value based on the track size */
  grid-template: "image" auto "content" 1fr "footer" auto / 100%;
}
```

The size of `.card-grid` does not accurately reflect the available space for a given card, but there is no other external "container" that `.card` can use to adjust the `grid-template`. Authors using this feature would need to add an extra wrapping element -- so that the card component has an external track-sized container to query:

```html
<section class="card-grid">
  <div class="card-container"><div class="card">...</div></div>
  <div class="card-container"><div class="card">...</div></div>
  <div class="card-container"><div class="card">...</div></div>
  <div class="card-container"><div class="card">...</div></div>
</section>
```

```css
/* the outer element can get containment… */
.card-container {
  contain: layout inline-size;
}

/* which gives .card something to query against */
@container (width > 30em) {
  .card {
    grid-template: "image content" 1fr "image footer" auto / 1fr 3fr;
  }
}
```

There are already many similar situations in CSS layout, so this might be a viable solution for most use-cases -- but the extra markup is not ideal.

### Single-axis containment issues

The existing [issue thread](https://github.com/w3c/csswg-drafts/issues/1031) has a proposal for moving forward with single-axis containment despite the known issues.

### Implicit vs explicit containers

In conversations leading to this proposal, there has been some concern about the dangers of establishing context _implicitly_ based on the value of `contain`. Similar behavior for positioning & stacking has sometimes been confusing for authors.

[David Baron's proposal][dbaron-cq] included a selector for querying a container more explicitly:

```css
/* syntax */
@container <selector> (<container-media-query>)? {
  /* ... */
}

/* example */
@container .media-object (width > 45em) {
  .media-object {
    grid-template: "img content" auto / auto 1fr;
  }
}
```

Since all known use-cases attempt to query the _most immediate available space_, I don't see any need for querying containers with an explicit syntax, or any way to "skipping over" one container to query the next.

Adding a selector to the query would also raise new problems:

- Explicitly targeted queries are less modular, so components would not be able to query _whatever contasiner they happen to be in_.
- It adds potential confusion about what selectors are allowed in the block. Authors would not be able to style the container itself, unless we limited the properties allowed -- similar to the `switch()` proposal.

However, it might be helpful to consider a more explicit way of defining the containers initially, to make this more clear for authors -- such as `query`, `inline-query`, & `block-query` values that would apply both layout and size containment. This needs more discussion & consideration.

### Combining scope with container queries?

David Baron's proposal also uses the explicit container selector to attach the concept of `scope` to containers -- only matching selectors inside the query against a subtree of the DOM. This might be useful for use-cases where a component both:

- Establishes its own containment context, and
- Establishes its own selector scope

But in my exploration of use-cases, it seems common that components will want to query _external_ context, while establishing _internal_ scope. There is also a mis-match where authors expect to style the root element of a given scope, but should not be able to style the root of a container-query. For those reasons, I think the two features -- container queries and scope -- should remain distinct, and be addressed separately.

### @-Rule or pseudo-class?

Many proposals & Javascript implementations use a pseudo-class rather than an @-rule.

```css
/* pseudo-class */
.selector:container(<query >) {
  /* ... */
}
```

I think the @-rule block provides several advantages:

- The @-rule syntax matches more closely with existing conditional rules, and builds on existing query-list syntax.
- It's likely that a responsive component will have multiple moving parts, and each might require unique selectors based on the same query. These can be grouped in an @-rule.
- We avoid the issues mentioned above with having an explicit selector attached to the query.

I also think the syntax can lead to confusion. It's not immediately clear what these different selectors would mean:

```css
:container(width < 40em) {
  font-size: small;
}
.media-object:container(width < 40em) {
  font-size: small;
}
:container(width < 40em) .media-object {
  font-size: small;
}
```

### Questions to explore

- Is it possible to apply containment within a container query? In what cases might that lead to cyclic dependencies?
- Will a default root fallback container feel like expected behavior, or should we require explicit containment in order to match queries?
- Are there special considerations required for a shadow-DOM `@container` to resolve queries against contained host element?
- Do we need a more explicit value for establishing query context, apart from existing `contain` values?
- What features can we support besides dimension-queries?

## References & acknowledgements

This proposal is based on the previous work of many people:

- Brian Kardell: [All Them Switches][switch]
- David Baron: [Thoughts on an implementable path forward][dbaron-cq]
- Mat Marquis: [A rough proposal for syntax](https://github.com/WICG/container-queries/issues/2)
- Matthew Dean: [2019 Proposal/Solution for Container Queries](https://github.com/WICG/container-queries/issues/12)
- Viktor Hubert: [Container Query Plugin](https://github.com/ZeeCoder/container-query/blob/master/docs/syntax.md#Queries)
- WICG: [Use Cases and Requirements](https://wicg.github.io/cq-usecases/)
- And more: [Who is Working on Container Queries](http://whoisworkingoncontainerqueries.com/)

Thanks also for valuable feedback and advice from:

- Adam Argyle
- Amelia Bellamy-Royds
- Anders Hartvoll Ruud
- Chris Coyier
- Christopher Kirk-Nielsen
- Eric Portis
- Ethan Marcotte
- Florian Rivoal
- Geoff Graham
- Gregory Wild-Smith
- Ian Kilpatrick
- Jen Simmons
- Martin Auswöger
- Martine Dowden
- Mike Riethmuller
- Morten Stenshorne
- Nicole Sullivan
- Rune Lillesveen
- Scott Jehl
- Scott Kellum
- Tab Atkins
- Theresa O’Connor
- Una Kravets

(sorry if I missed anyone… I don't mean to imply all these people has _signed off_ on my proposal, or even seen the whole thing - but they have all provided useful feedback along the way)

**Comment by mirisuzanne** [comment]
There has been some useful feedback in various channels, so I've been updating and maintaining a full [explainer with a changelog](https://github.com/oddbird/css-sandbox/blob/main/src/rwd/query/explainer.md) in my own repo. It's mostly minor clarifications at this point, but I wanted to make sure the latest updates are available from this thread as well.

**Comment by matthew-dean** [comment]
@mirisuzanne THANK YOU so much for this, and thank you for acknowledging my small contribution! I'm really excited about this and hope it happens! Love the at-rule syntax! ❤️

**Comment by matthew-dean** [comment]
I've been thinking about this proposal a lot since I came across it, and I just want to mention something.

When you're querying the viewport with `@media`, there's essentially a single "object" being queries. While I totally get the argument that with this `@container` syntax, a selector is not needed, I believe consideration should be given to the fact that _omitting_ a selector, even a star selector, (IMO) makes this conceptually harder to understand for new CSS authors. "What am I querying?" "Well it's the available container dimensions anywhere." " 🤔  "

If you're intimately familiar with CSS and browser layout mechanisms, it's not hard to grok a container query without a selector. But for new authors, I feel like the ability to query a selector should at least be _optional_, if not recommended.

e.g.
```css
@container * (width > 45em) {
  .media-object {
    grid-template: "img content" auto / auto 1fr;
  }
}
```
To me, this is an easier mental model, and there's good reason it's been included in almost every container query proposal. It ties into the mental model for CSS authors the "what" of what I'm querying, followed by the query, and IMO is the most analogous to a `@media` query, in which the "what" is clearly established, not explicitly in the syntax, but implicitly by the nature of the query.

I wouldn't consider this a feature blocker, of course, because I'm so jazzed that this proposal has legs, and honored to have been mentioned, and for me personally, I won't struggle with the mental model. But the more I think about it, the more I worry that the omission of a selector as even an option makes this hard for new devs, because the "what" takes a lot of concept knowledge. Couple that with needing to understand the `contain` property, and it's a lot to grok.

As to these points arguing against the inclusion of a selector from Dave Baron's proposal:

> Explicitly targeted queries are less modular, so components would not be able to query whatever contasiner they happen to be in.

That may be exactly what the author desires! An author may want to write queries which only apply in specific contexts (and change in others.) And to make them ultra-modular, one would target a `*` selector, which is easy for new authors to understand, because it ties into the basics of CSS.

> It adds potential confusion about what selectors are allowed in the block. 

To me, the inverse is true. A selector, even a star selector, makes the context of the query clear. Omitting a selector in the query muddies the "what", leading to more possible developer confusion. For advanced authors, the option to omit a selector makes a certain sense; you (usually) want this to apply regardless of context. But even for advanced authors, I can see myself wanting to apply a query only within certain contexts, and it makes the feature much more powerful as a result.

There are a lot of very talented people / CSS experts who've weighed into this proposal. My suggestion is just to consider the non-expert. Will they understand the context? What syntax is most intuitive? Have they been asked? I can imagine a person teaching this CSS concept to someone new and saying something like, "Imagine there's a star selector in this query, even though it's not present." That lesson would be made easier if the selector was _actually_ present.

Okay, that's enough of my $0.02. Thanks so much for moving this along!

**Comment by cssinate** [comment]
@matthew-dean 

MAYBE I'm mistaken, but if you do want to scope it, don't you do something like:

```css
@contain (width > 45em) {
  .theScopeIWant .elementSelected {
    ...
  }
}
```

**Comment by maxnordlund** [comment]
I love this. Gonna make so much stuff easy and beautiful. I think that keeping a selector out of it makes sense, out keeps the syntax similar to `@media`. Having `contain` create a container context feels like `position: relative` / `position: absolute`, or `z-index`ing. I also like how it sidesteps styling the container itself, which could lead to infinite loops.

❤️

**Comment by matthew-dean** [comment]
@cssinate I don't believe so. The container scope should be the immediate parent that you're querying. Everything inside is within that queried container, which is why I believe this syntax is potentially confusing.

**Comment by mirisuzanne** [comment]
My slides are archived here as a PDF, for anyone interested: https://lists.w3.org/Archives/Public/www-archive/2021Feb/att-0002/Container_Queries_Proposal_-_vF2F_2021-02-09.pdf

> Everything inside is within that queried container

@matthew-dean Not exactly. @cssinate was right about this. The entire selector is not scoped to a container - only the target element. My goal is for _containment context_ to match _positioning context_. So you can do:

```css
.theScopeIWant { contain: <something>; }

@container (width > 45em) {
  .theScopeIWant .elementSelected { ... }
}
```

I'm happy to discuss that more, and provide more of my reasoning. In the end it might even be possible to support both. But I think at this point it would be best handled in a dedicated issue. Do you want to open that issue, and link back to this conversation? (if not, I'll get to it soon)

**Comment by MostafaLogic2002** [comment]
Nice work❤️❤️!..Keep Going

**Comment by matthew-dean** [comment]
@mirisuzanne Thanks for the clarification. I think.... I may be misunderstanding what @cssinate means by scope, or a fundamental piece of this proposal.

I guess maybe what you're saying is that you can limit the scope of a query by placing the scope within the realm of the `@container` query.

So, I think what you're saying is that you don't _need_ a selector because you can already say, via the child selector within `@container`, to only apply the query based on, say, the fact that `.elementSelected` is a descendent of `.theScopeIWant`? 🤔  i.e. given this CSS:
```css
@container (width > 45em) {
  .theScopeIWant .elementSelected {
    ...
  }
}
```
One of the things I tried to illustrate with my proposal is that everyone thinks about this problem very differently! If I misunderstood some of the fundamentals of your proposal, I apologize! I honestly feel this is a far cleaner syntax than what I proposed, but it seems my fear that someone wouldn't get the mental model right should have been self-applied lol.

So to clarify, this:
```css
@container (width > 45em) {
  .card { ... }
  h2 { ... }
}
```
is essentially similar to:
```css
.card:container(width > 45em) { ... }
h2:container(width > 45em) { ... }
```
In that, it applies to`.card` and `h2` given that they would be rendering in a space (or container?) that is > 45em?

I think this is such a departure from previous proposals that maybe I didn't quite get it. That being the case, I'm not sure another issue needs to be raised because I'm not sure there is one? If I misunderstood (and am understanding it better now?) then I'm happy to remove my previous post so that I don't confuse anyone else.

Thanks for taking the time to respond!

**Comment by mirisuzanne** [comment]
@matthew-dean yep, you got it. Exactly. No need to remove anything - these are exactly the conversations we need to be having, and help clarify the language. I'm glad for your input.

(Issues are free, and don't have to represent a problem – just a decision that needs to be made, or a question that needs more clarity. I'll soon start opening a number of them against my own proposal, as I start writing the spec. Feel free to do the same if you (_anyone reading this_) have other questions, or followup. Issues can be labeled with `[css-contain-3]` for clarity.)

**Comment by tomhodgins** [comment]
This is fantastic news! Congrats everyone, and thanks CSSWG! 😍

**Comment by tombigel** [comment]
A thought about scoped container queries (I hope I didn't miss something written about this before)

The first thing I thought about when reading the scoped container section was the `@keyframes` syntax - why not name queries so they are reusable instead of limiting ourselves to specific selectors? 
Media queries has one scope - the media, but container queries should have the option to relate to more than one scope:

Something like -
```css
@container someNamedQuery (width > 45em) {
  ...
}

@container otherNamedQuery (width > 80em) {
  ...
}

/* All sections can have one rule (don't mind my naming, I'm bad at it) */
section {
  contain: layout size;
  container-query: someNamedQuery;
}

/* But we override this rule for sections that should behave differently */
section.special-rules {
  contain: layout size;
  container-query: otherNamedQuery;
}
```

**Comment by kospl** [comment]
I know you probably discussed this situation, but anyway

```
@container .media-object (width > 45em) {
  .media-object {
    width: 1px;
  }
}
```

**Comment by wUFr** [comment]
Hello, im probably late for the party, but i maybe would like something like this:

```css
.sidebar {
	width: 100%;
	layout: mobile; /* use whatever name you want, these are examples; passed to every child element */
}

@media (min-width:900px){
	.sidebar {
		width: 200px;
		layout: desktop;
	}
}

@media (min-width:1600px){
	.sidebar {
		width: 400px;
		layout: desktop-wide
	}
}

.sidebar > ul:layout(mobile){
	/* styles for mobile */
}

.sidebar > ul:layout(desktop){
	/* styles for desktop */
}


.sidebar > ul:layout(desktop-wide){
	/* styles for wide desktop */
}
```

the "layout" variable would be available to every child element, so it would be easy to access in case a lot of nested code. Example from less:
```less
.sidebar {
	...

	ul {
		...

		li {
			a {
				span:layout(mobile){
					display: none;
				}

				span:layout(desktop-wide){
					display: block;
				}
			}
		}
	}
}
```

this way containing elements would use name instead of dynamic values, which could be changed at one spot (container settings, sidebar in this case). So in case i would decided i want "desktop" view starting from 1000px not 900px i would change it only once. I know this could be probably achieved by using variables via pre-processors (or maybe native?), but this looks a bit cleaner to me :)

**Comment by nhoizey** [comment]
@wUFr what you suggest looks a lot like viewport wide media queries instead of container queries.

If you replace `.sidebar > ul:layout(desktop)` with `@media (min-width:900px){ .sidebar > ul { … } }`, there is no need for container queries.

And I think you can mix it with custom properties to pass context to child elements, and use it with this nice trick from Lea Verou: https://lea.verou.me/2020/10/the-var-space-hack-to-toggle-multiple-values-with-one-custom-property/

**Comment by wUFr** [comment]
@nhoizey thinking about it, you're right, but it would be nice to have something like this too :beer:

**Comment by nhoizey** [comment]
@wUFr I agree, you could open another issue for this.

**Comment by matthew-dean** [comment]
Thanks @mirisuzanne for continuing to move this forward!

**Comment by MajPay** [comment]
I would absolutely love to see native container queries realized. One suggestion though:

Please think about supporting css-vars inside the queries

Example:

```css
@container var(--container-medium, (width > 480px)) {
    content: "i am displayed inside medium sized containers";
}
```

This would solve many issues mentioned above and add further flexibility. You could provide a component/framework/whatever with good defaults but also support custom values in a very easy way and the base technology for this (css-vars) is already available!

integration example (lets assume loading a web-component, that uses container queries):

```html
<script type="module" src="//some.cdn/some-component.js"></script>
<style type="text/css">
:root {
  --container-medium: (width > 520px);
}
</style>
```

**Comment by matthew-dean** [comment]
@MajPay That wouldn't work in practice. A `var()` only has a value through the cascade. At the level of `@container`, it has no value, because `--container-medium` is only a value within `:root` within your stylesheet. So your usage is more like Less / Sass's uses of variables within a stylesheet, which CSS doesn't yet support (but hopefully will someday with being able to set global env vars).

**Comment by mirisuzanne** [comment]
> Please think about supporting css-vars inside the queries

While this exact solution isn't possible, there has been [some discussion](https://github.com/w3c/csswg-drafts/issues/5009#issuecomment-620726494) of adding a new syntax for custom reusable media-queries. If something like that happens, I expect we would now want it to support container queries as well. But that's a separate feature/conversation.
